### PR TITLE
`destroy` testing modules during `resetTestingModule`

### LINKED ISSUE #18831: Destroy listeners aren't called in TestBed
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
Destroy listeners that are registered on `PlatformRef` and `NgModuleRef` with `onDestroy` aren't called on test bed teardown.

## Expected behavior
It's expected that destroy listeners will be called automatically.

## Minimal reproduction of the problem with instructions
A [plunk](http://plnkr.co/edit/kYwrjgnTa7QGwLDGjQcO?p=preview):

```
@NgModule()
export class FooModule {
  constructor(moduleRef: NgModuleRef, platformRef: PlatformRef) {
    moduleRef.onDestroy(() => { throw 'module destroy' });
    platformRef.onDestroy(() => { throw 'platform destroy' });
  }
}

describe('tests', () => {
  beforeEach(() => {
    TestBed.resetTestEnvironment(); 
    TestBed.initTestEnvironment(
      BrowserDynamicTestingModule,
      platformBrowserDynamicTesting()
    );
  });

  beforeEach(() => {
    TestBed.configureTestingModule({ imports: [FooModule] });
  });
 
  it('test', inject([], () => {
    expect(1).toBe(1)
  });
});
```

## What is the motivation / use case for changing the behavior?

Modules can have cleanup code that should be executed in order for them to function correctly in environments with multiple bootstraps (which TestBed tests are).

## Environment

<pre><code>
Angular version: 4.3.5

### PR DESCRIPTION
**Pull request by josephperrott**
See individual commits.

Fixes #18831

### CODE REVIEW

**Comment by muhammedMoussa** [comment]
Hi, can `(this.subscription && !this.subscription.closed)` shorthanded <br> to be ``` (!this.subscription?.closed)```

File: packages/platform-server/src/server_renderer.ts [hunk 1]
```diff
@@ -249,9 +251,10 @@ class EmulatedEncapsulationServerRenderer2 extends DefaultServerRenderer2 {
   private hostAttr: string;
 
   constructor(
-      eventManager: EventManager, document: any, ngZone: NgZone, sharedStylesHost: SharedStylesHost,
+      eventManager: EventManager, document: any, ngZone: NgZone, sharedStylesHost: ServerStylesHost,
       schema: DomElementSchemaRegistry, private component: RendererType2) {
     super(eventManager, document, ngZone, schema);
+    console.log('renderer created');
```
Comment (Splaktar) [inline packages/platform-server/src/server_renderer.ts:257]: Was this just for debugging? Can it be removed now?

**Review by Splaktar — CHANGES_REQUESTED** [review]
It looks like CircleCI is requiring this PR be rebased:
```
        CircleCI config on master has been modified since commit 250e299,
        which this PR is based on.

        Please rebase the PR on master after fetching from upstream.

        Rebase instructions for PR Author, please run the following commands:

          git fetch upstream master;
          git checkout shared-styles-cleanup;
          git rebase upstream/master;
          git push --force-with-lease;
```

**Comment by elesueur** [comment]
@AndrewKushnir can you describe how this change is breaking?

**Comment by AndrewKushnir** [comment]
@elesueur given that there was no cleanup of shared styles, tests might be written in a way where they depend on the presence of some CSS rules added during previous tests. With the proposed cleanup, such tests will start to fail (so this is a breaking change). We've confirmed that by running tests in Google's codebase. Before merging this change it's important to estimate potential risks and come up with proper rollout plan/communication.

@josephperrott please correct me if I'm wrong or if I'm missing something.

**Comment by josephperrott** [comment]
That is correct, I would only add that it could go even further and maintain the state in other ways between tests.  Since modules are not torn down, things like the state of a service can leak between tests.

**Comment by elesueur** [comment]
> That is correct, I would only add that it could go even further and maintain the state in other ways between tests. Since modules are not torn down, things like the state of a service can leak between tests.

As a maintainer of several large Angular apps, I would want to learn about which tests are relying on pollution (and fix them) and get the performance improvement rather than wait for a 'less breaky' solution that hides the pollution and leaves me with slow tests :)

**Comment by crisbeto** [comment]
Will be superseded by https://github.com/angular/angular/pull/42566.

**Comment by josephperrott** [comment]
Closing in favor of #42566

**Pull request closed without merging**